### PR TITLE
Finetune guide for semantic segmentation

### PR DESCRIPTION
**Pull request by stevhliu**
This PR creates a finetune guide for semantic segmentation in the docs. Unlike previous finetune guides, this one will include:

* metrics for evaluation
* a section for how to use the model for inference
* an embedded Gradio demo

🚧 To do:
- [x] create section for inference
- [ ] create Gradio demo

### CODE REVIEW

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# Semantic segmentation
+
+<Youtube id="dKE8SIt9C-w"/>
+
+Semantic segmentation assigns a label or class to each individual pixel of an image. There are several types of segmentation tasks, and in the case of semantic segmentation, it doesn't differentiate between unique instances of the same object. Both objects are given the same label. Common real-world applications of semantic segmentation include training self-driving cars to identify pedestrians and important traffic information, identifying cells and abnormalities in medical imagery, and monitoring environmental changes from satellite imagery.
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:17]:
```suggestion
Semantic segmentation assigns a label or class to each individual pixel of an image. There are several types of segmentation, and in the case of semantic segmentation, one doesn't differentiate between unique instances of the same object. Both objects are given the same label (e.g. "car"). Common real-world applications of semantic segmentation include training self-driving cars to identify pedestrians and important traffic information, identifying cells and abnormalities in medical imagery, and monitoring environmental changes from satellite imagery.
```

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+>>> repo_id = "datasets/huggingface/label-files"
+>>> filename = "ade20k-id2label.json"
+>>> id2label = json.load(open(cached_download(hf_hub_url(repo_id, filename)), "r"))
+>>> id2label = {int(k): v for k, v in id2label.items()}
+>>> label2id = {v: k for k, v in id2label.items()}
+```
+
+## Preprocess
+
+Next, load a SegFormer feature extractor to process the image into an array of tensors. Some datasets, like this one, uses the zero-index as the background class. However, the background class isn't included in the 150 classes, so you'll need to set `reduce_labels=True` to subtract one from all the labels. The zero-index is replaced by `255` so it's ignored by SegFormer's loss function:
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:77]: Note that we're currently in the process of replacing feature extractors by image processors, as feature extractor is a confusing name.

Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:77]:
```suggestion
Next, load a SegFormer feature extractor to process the image into an array of tensors. Some datasets, like this one, use the zero-index as the background class. However, the background class isn't included in the 150 classes, so you'll need to set `reduce_labels=True` to subtract one from all the labels. The zero-index is replaced by `255` so it's ignored by SegFormer's loss function:
```

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+It is common to apply some data augmentations to an image dataset to make a model more robust against overfitting. In this guide, you'll use the [`ColorJitter`](https://pytorch.org/vision/stable/generated/torchvision.transforms.ColorJitter.html) function from [torchvision](https://pytorch.org/vision/stable/index.html) to randomly change the color properties of an image:
+
+```py
+>>> from torchvision.transforms import ColorJitter
+
+>>> jitter = ColorJitter(brightness=0.25, contrast=0.25, saturation=0.25, hue=0.1)
+```
+
+Now create a preprocessing function that'll apply the `jitter` transform and return the `pixel_values` and `labels` of an image. `pixel_values` is the expected model input and it is generated by the feature extractor. You should only `jitter` the images in the training set. For the `labels` and the test set, the feature extractor will only crop and normalize the `images` and `labels`.
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:93]:
```suggestion
Now let's create a preprocessing function that will apply the `jitter` transform and return the `pixel_values` and `labels` of an image. `pixel_values` is the expected model input and it is generated by the feature extractor. You should only `jitter` the images in the training set. For the `labels` and the test set, the feature extractor will only crop and normalize the `images` and `labels`.
```

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+<pt>
+Load SegFormer with [`AutoModelForSemanticSegmentation`], specify the number of labels, and pass the model the mapping between label ids and label classes:
+
+```py
+>>> from transformers import AutoModelForSemanticSegmentation
+
+>>> pretrained_model_name = "nvidia/mit-b0"
+>>> model = AutoModelForSemanticSegmentation.from_pretrained(
+...     pretrained_model_name, num_labels=num_labels, id2label=id2label, label2id=label2id
+... )
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:129]: @sgugger told me one shouldn't provide the `num_labels` argument in case you provide `id2label` mappings

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+```
+
+<Tip>
+
+If you aren't familiar with finetuning a model with the [`Trainer`], take a look at the basic tutorial [here](../training#finetune-with-trainer)!
+
+</Tip>
+
+Define your training hyperparameters in [`TrainingArguments`]. It is important you don't remove unused columns because this will drop the `image` column. Without the `image` column, you can't create `pixel_values`. Set `remove_unused_columns=False` to prevent this behavior!
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:138]:
```suggestion
Define your training hyperparameters in [`TrainingArguments`]. It is important that you don't remove unused columns because this will drop the `image` column. Without the `image` column, you can't create `pixel_values`. Set `remove_unused_columns=False` to prevent this behavior!
```

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+...     save_strategy="steps",
+...     save_steps=20,
+...     eval_steps=20,
+...     logging_steps=1,
+...     eval_accumulation_steps=5,
+...     remove_unused_columns=False,
+... )
+```
+
+To evaluate model performance during training, you'll need to create a function to compute and report metrics. For semantic segmentation, you'll typically compute the [mean intersection over union](https://huggingface.co/spaces/evaluate-metric/mean_iou) (IoU). The mean IoU measures the overlapping area between the predicted and ground truth segmentation maps.
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:160]:
```suggestion
To evaluate model performance during training, you'll need to create a function to compute and report metrics. For semantic segmentation, you'll typically compute the [mean Intersection over Union](https://huggingface.co/spaces/evaluate-metric/mean_iou) (IoU). The mean IoU measures the overlapping area between the predicted and ground truth segmentation maps. 
```

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+Then take a look at an example:
+
+```py
+>>> train_ds[0]
+{'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=512x683 at 0x7F9B0C201F90>,
+ 'annotation': <PIL.PngImagePlugin.PngImageFile image mode=L size=512x683 at 0x7F9B0C201DD0>,
+ 'scene_category': 368}
+```
+
+There is an `image`, an `annotation` (this is the segmentation map or label), and a `scene_category` field that describes the image scene like kitchen or office. Both `image` and `annotation` are PIL images.
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:60]:
```suggestion
There is an `image`, an `annotation` (this is the segmentation map or label), and a `scene_category` field that describes the image scene, like "kitchen" or "office". We'll only need `image` and `annotation`, both are PIL images.
```

File: docs/source/en/tasks/semantic_segmentation.mdx (hunk 2 above)
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:77]:
```suggestion
Next, load a SegFormer feature extractor, which can be used to prepare the images and annotations for the model. Some datasets, like this one, use the zero-index as the background class. However, the background class isn't included in the 150 classes, so you'll need to set `reduce_labels=True` to subtract one from all the labels. The zero-index is replaced by `255` so it's ignored by SegFormer's loss function:
```

File: docs/source/en/tasks/semantic_segmentation.mdx (hunk 4 above)
Comment (NielsRogge) [reply docs/source/en/tasks/semantic_segmentation.mdx:129]: Not sure though whether that works

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+If you aren't familiar with finetuning a model with the [`Trainer`], take a look at the basic tutorial [here](../training#finetune-with-trainer)!
+
+</Tip>
+
+Define your training hyperparameters in [`TrainingArguments`]. It is important you don't remove unused columns because this will drop the `image` column. Without the `image` column, you can't create `pixel_values`. Set `remove_unused_columns=False` to prevent this behavior!
+
+```py
+>>> from transformers import TrainingArguments
+
+>>> training_args = TrainingArguments(
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:143]: I'd also add push to hub here

File: docs/source/en/tasks/semantic_segmentation.mdx (hunk 2 above)
Comment (stevhliu) [reply docs/source/en/tasks/semantic_segmentation.mdx:77]: Thanks for the heads up, I'll update this again once the image processors are ready

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+```py
+>>> train_ds.set_transform(train_transforms)
+>>> test_ds.set_transform(val_transforms)
+```
+
+## Train
+
+<frameworkcontent>
+<pt>
+Load SegFormer with [`AutoModelForSemanticSegmentation`], specify the number of labels, and pass the model the mapping between label ids and label classes:
```
[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/tasks/semantic_segmentation.mdx:121]: I think since we have a Tip below that links to the finetuning tutorial where this is explained, it's not necessary to add this clarification.

File: docs/source/en/tasks/semantic_segmentation.mdx (hunk 4 above)
Comment (stevhliu) [reply docs/source/en/tasks/semantic_segmentation.mdx:129]: Oh, interesting! I was able to finetune without `num_labels`, but we'll wait for more clarification from Sylvain :)

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+```py
+>>> train_ds.set_transform(train_transforms)
+>>> test_ds.set_transform(val_transforms)
+```
+
+## Train
+
+<frameworkcontent>
+<pt>
```
Comment (sgugger) [inline docs/source/en/tasks/semantic_segmentation.mdx:120]: The whole tutorial is PyTorch only (it uses torchvision before), so not sure this is useful here.

File: docs/source/en/tasks/semantic_segmentation.mdx (hunk 4 above)
Comment (sgugger) [reply docs/source/en/tasks/semantic_segmentation.mdx:129]: Yes, you don't need to pass `num_labels` if you pass `id2label`, it will just be ignored.

**Review by sgugger — APPROVED** [review]
Nice addition!

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# Semantic segmentation
+
+<Youtube id="dKE8SIt9C-w"/>
+
+Semantic segmentation assigns a label or class to each individual pixel of an image. There are several types of segmentation, and in the case of semantic segmentation, it doesn't differentiate between unique instances of the same object. Both objects are given the same label (for example, "car" instead of "car-1" and "car-2"). Common real-world applications of semantic segmentation include training self-driving cars to identify pedestrians and important traffic information, identifying cells and abnormalities in medical imagery, and monitoring environmental changes from satellite imagery.
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:17]:
```suggestion
Semantic segmentation assigns a label or class to each individual pixel of an image. There are several types of segmentation, and in the case of semantic segmentation, no distinction is made between unique instances of the same object. Both objects are given the same label (for example, "car" instead of "car-1" and "car-2"). Common real-world applications of semantic segmentation include training self-driving cars to identify pedestrians and important traffic information, identifying cells and abnormalities in medical imagery, and monitoring environmental changes from satellite imagery.
```

File: docs/source/en/tasks/semantic_segmentation.mdx [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+
+
+>>> def val_transforms(example_batch):
+...     images = [x for x in example_batch["image"]]
+...     labels = [x for x in example_batch["annotation"]]
+...     inputs = feature_extractor(images, labels)
+...     return inputs
+```
+
+To apply the `jitter` over the entire dataset, use 🤗 Datasets [`~datasets.Dataset.set_transform`] function. The transform is applied on the fly which is faster and consumes less disk space:
```
Comment (NielsRogge) [inline docs/source/en/tasks/semantic_segmentation.mdx:111]:
```suggestion
To apply the `jitter` over the entire dataset, use the 🤗 Datasets [`~datasets.Dataset.set_transform`] function. The transform is applied on the fly which is faster and consumes less disk space:
```

**Review by NielsRogge — APPROVED** [review]
Thanks for adding this!
